File: src/gui/mrview/tool/fixel/legacy.cpp
/* Copyright (c) 2008-2017 the MRtrix3 contributors
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, you can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * MRtrix is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty
 * of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 *
 * For more details, see http://www.mrtrix.org/.
 */


#include "gui/mrview/tool/fixel/legacy.h"

namespace MR
{
  namespace GUI
  {
    namespace MRView
    {
      namespace Tool
      {

        void Legacy::load_image_buffer()
        {
          for (size_t axis = 0; axis < 3; ++axis) {
            const size_t axis_size = fixel_data->size (axis);
            slice_fixel_indices[axis].resize (axis_size);
            slice_fixel_sizes  [axis].resize (axis_size);
            slice_fixel_counts [axis].resize (axis_size, 0);
          }

          FixelValue& fixel_val_store = fixel_values[value_types[1]];
          FixelValue& fixel_secondary_val_store = fixel_values[value_types[2]];

          for (auto l = Loop (*fixel_data) (*fixel_data); l; ++l) {

            const std::array<int, 3> voxel {{ int(fixel_data->index (0)),
              int(fixel_data->index (1)), int(fixel_data->index (2)) }};

            Eigen::Vector3f pos { float(voxel[0]), float(voxel[1]), float(voxel[2]) };
            pos = transform.voxel2scanner.cast<float> () * pos;

            for (size_t f = 0; f != fixel_data->value ().size (); ++f) {

              auto val = fixel_data->value ()[f].size;
              auto secondary_val = fixel_data->value ()[f].value;

              pos_buffer_store.push_back (pos);
              dir_buffer_store.push_back (fixel_data->value ()[f].dir);

              fixel_val_store.add_value (val);
              fixel_secondary_val_store.add_value (secondary_val);

              GLint point_index = pos_buffer_store.size () - 1;

              for (size_t axis = 0; axis < 3; ++axis) {
                slice_fixel_indices[axis][voxel[axis]].push_back (point_index);
                slice_fixel_sizes  [axis][voxel[axis]].push_back (1);
                slice_fixel_counts [axis][voxel[axis]]++;
              }

              voxel_to_indices_map[voxel].push_back (point_index);
            }
          }
        }
      }
    }
  }
}

File: python/scripts/dwi2mask/legacy.py
# Copyright (c) 2008-2023 the MRtrix3 contributors.
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
#
# Covered Software is provided under this License on an "as is"
# basis, without warranty of any kind, either expressed, implied, or
# statutory, including, without limitation, warranties that the
# Covered Software is free of defects, merchantable, fit for a
# particular purpose or non-infringing.
# See the Mozilla Public License v. 2.0 for more details.
#
# For more details, see http://www.mrtrix.org/.

from mrtrix3 import app, run

DEFAULT_CLEAN_SCALE = 2



def usage(base_parser, subparsers): #pylint: disable=unused-variable
  parser = subparsers.add_parser('legacy', parents=[base_parser])
  parser.set_author('Robert E. Smith (robert.smith@florey.edu.au)')
  parser.set_synopsis('Use the legacy MRtrix3 dwi2mask heuristic (based on thresholded trace images)')
  parser.add_argument('input',  help='The input DWI series')
  parser.add_argument('output', help='The output mask image')
  parser.add_argument('-clean_scale',
                      type=int,
                      default=DEFAULT_CLEAN_SCALE,
                      help='the maximum scale used to cut bridges. A certain maximum scale cuts '
                           'bridges up to a width (in voxels) of 2x the provided scale. Setting '
                           'this to 0 disables the mask cleaning step. (Default: ' + str(DEFAULT_CLEAN_SCALE) + ')')



def get_inputs(): #pylint: disable=unused-variable
  pass



def needs_mean_bzero(): #pylint: disable=unused-variable
  return False



def execute(): #pylint: disable=unused-variable

  run.command('mrcalc input.mif 0 -max input_nonneg.mif')
  run.command('dwishellmath input_nonneg.mif mean trace.mif')
  app.cleanup('input_nonneg.mif')
  run.command('mrthreshold trace.mif - -comparison gt | '
              'mrmath - max -axis 3 - | '
              'maskfilter - median - | '
              'maskfilter - connect -largest - | '
              'maskfilter - fill - | '
              'maskfilter - clean -scale ' + str(app.ARGS.clean_scale) + ' mask.mif')

  return 'mask.mif'
